File: src/main/python/HAnim.py
import xml.etree.ElementTree as ET

# list Joint centers in Katy's format

tree = ET.parse("../../specifications/X3dUnifiedObjectModel-4.0.xml")

root = tree.getroot()

sites = root.findall(".//SimpleType[@name='hanimFeaturePointNameValues']")
joints = root.findall(".//SimpleType[@name='hanimJointNameValues']")
segments = root.findall(".//SimpleType[@name='hanimSegmentNameValues']")
#versions = root.findall(".//SimpleType[@name='hanimVersionChoices']")

print("--------------------------------")
print("Joints")
for elem in joints:
    # print(ET.tostring(elem))
    enums = elem.findall(".//enumeration")
    for enum in enums:
        # print(ET.tostring(enum))
        print("--------------------------------")
        print(enum.get('value'))
        print("Type:", "joint")
        print("Parent:", enum.get('parent'))
        print("Index:", enum.get('index'))
        xyz = enum.get('default')
        # print(xyz)
        if xyz is not None:
            default = xyz.split(" ")
            print("X:", default[0], "m")
            print("Y:", default[1], "m")
            print("Z:", default[2], "m")
        else:
            print("X:", "x", "m")
            print("Y:", "y", "m")
            print("Z:", "z", "m")
        if enum.get('alias') is not None:
            print("Aliases:", enum.get('alias'))
        print("")

print("--------------------------------")
print("Sites")
for elem in sites:
    # print(ET.tostring(elem))
    enums = elem.findall(".//enumeration")
    for enum in enums:
        #print(ET.tostring(enum))
        print("--------------------------------")
        print(enum.get('value'))
        print("Type:", "site")
        print("Parent:", enum.get('parent'))
        print("Index:", enum.get('index'))
        xyz = enum.get('default')
        if xyz is not None:
            default = xyz.split(" ")
            # print(default)
            print("X:", default[0], "m")
            print("Y:", default[1], "m")
            print("Z:", default[2], "m")
        else:
            print("X:", "x", "m")
            print("Y:", "y", "m")
            print("Z:", "z", "m")
        if enum.get('alias') is not None:
            print("Aliases:", enum.get('alias'))
        print("")

print("--------------------------------")
print("Segments")
for elem in segments:
    # print(ET.tostring(elem))
    enums = elem.findall(".//enumeration[@value]")
    for enum in enums:
        # print(ET.tostring(enum))
        print("--------------------------------")
        print(enum.get('value'))
        print("Parent:", enum.get('parent'))
        print("Type:", "segment")
        print("Index:", enum.get('index'))
        if enum.get('alias') is not None:
            print("Aliases:", enum.get('alias'))
        print("")

File: src/main/node/hanim.js
var fs = require('fs');
skeleton = {}

let DOM = fs.readFileSync("../__/specifications/X3dUnifiedObjectModel_4_0.json");
let SimpleTypes = JSON.parse(DOM)["X3dUnifiedObjectModel"]['-children'][0]["SimpleTypeEnumerations"]['-children'];
// console.log(SimpleTypes)
for (let ST in SimpleTypes) {
	let SimpleType = SimpleTypes[ST]["SimpleType"]
	// console.log(SimpleType)
	if (typeof SimpleType !== 'undefined' && typeof SimpleType["@name"] !== 'undefined') {
		
		if (SimpleType["@name"] === "hanimJointNameValues" || SimpleType["@name"] === "hanimSegmentNameValues") {
			console.log(SimpleType["@name"])
			for (let enumerationsIndex in SimpleType["-children"]) {
				let enumeration = SimpleType["-children"][enumerationsIndex]["enumeration"];
				console.log(JSON.stringify(skeleton, null, 2))
				insertIntoTree(skeleton, enumeration['@parent'], enumeration['@value'])
			}
		}
	}
}

function insertIntoTree(tree, parent, child) {
	// console.log(tree, parent, child);
	console.log(JSON.stringify(tree, null, 2))
	if (Object.keys(tree).length === 0) {
		if (typeof parent !== 'undefined') {
			tree[parent] = []
		} else {
			tree[child] = [];
		}
	} else {
		for (n in tree) {
			if (isNaN(n)) {
				let node = tree[n];
				if (n !== parent) {
					//console.log(node, parent, child);
					insertIntoTree(node, parent, child);
				} else { // found
					//console.log(node, parent, child);
					tree[parent][child] = [];
					break;
					
				}						
			}
		}
	}
	return tree;
}
console.log(skeleton);
